### ISSUE TITLE
Attributes does not preserve order

### ISSUE THREAD

**Issue by bojanv55** [thymeleaf/thymeleaf#617]
When I use `th:attr="__${event.attribs}__"`, attributes are rendered in random order and not in order presented in event.attribs String. E.g. in string I have `data-sport='Tennis',data-association='ATP',data-country='France',data-gender='Men',data-tennis-match-type='Singles',data-tournament='French Open'` and output is `data-gender="Men" data-association="ATP" data-country="France" data-tennis-match-type="Singles" data-sport="Tennis" data-tournament="French Open"`

**Comment by danielfernandez** [comment]
HTML attributes are unordered by spec, though it is true that keeping a specific order could improve human (not machine) understanding of the code. So Thymeleaf 3.0 tries to respect order in many of its operations involving attributes… am I right therefore in assuming you are using an older version of Thymeleaf?

**Comment by bojanv55** [comment]
I'm aware of this spec, but I have some 3rd party js scripts are using this order as their presentation ordering. I fixed this in js source, but anyway wanted to report. I'm using current stable spring starter dependency.

**Comment by danielfernandez** [comment]
What version of Thymeleaf are you using?

**Comment by bojanv55** [comment]
Will check tomorrow (not @ pc now). Hoped you know which one is bundled in current stable starter spring boot.

**Comment by danielfernandez** [comment]
Yes, but I missed that *"starter"* in your previous comment. By default current the version of Spring Boot (1.x) uses Thymeleaf 2.1, though this [can be changed to use 3.0](https://docs.spring.io/spring-boot/docs/current/reference/html/howto-spring-mvc.html#howto-use-thymeleaf-3). Spring Boot 2.0 will use v3.0 by default.

Thymeleaf 2.1 does not pay attention to the order of attributes specified in `th:attr`. I will leave this open as a wishlist item for future maintenance versions of the 2.1 branch. Thymeleaf 3.0 does preserve order (the entire engine works in a completely different way), though it does this for human readability reasons as mentioned.

Anyway I'd recommend filing a bug with your 3rd party JavaScript library as IMHO relying on attribute order would definitely be a bug for a browser-side lib.

**Comment by danielfernandez** [comment]
This is no longer an issue in modern versions.
